File: 400_Nth_Digit/findNthDigit.cpp
#include <iostream>
#include <math.h>
#include <string>

using namespace std;

class Solution {
    public:
        int findNthDigit(int n) {
            /* calculate how maniy digits the number has. */
            long base = 9, digits = 1;
            while(n - base * digits > 0) {
                n -= base * digits;
                base *= 10;
                digits++;
            }

            /* calculate what the muber is */
            int index = n % digits;
            if(index == 0) index = digits;

            long num = 1;
            for(int i = 1; i < digits; i++)
                num *= 10;

            num += (index == digits)? n / digits - 1 : n / digits;

            int res = to_string(num)[index - 1] - '0';
            return res;
        }
};

int main() {
    Solution sol;
    cout << sol.findNthDigit(15) << endl;
    return 0;
}

File: 400_Nth_Digit/findNthDigit.py
class Solution(object):
    def findNthDigit(self, n):
        if n < 10: return n
        pow = 0
        tmp = n
        while tmp > 0:
            tmp -= (9 * (10 ** pow)) * (pow + 1)
            pow += 1
        
        Nth = tmp + (9 * (10 ** (pow - 1))) * pow
        #print "Nth, pow: ", Nth, pow
        # pow digits, the Nth digit
        v = Nth / pow
        count = v if Nth % pow == 0 else v + 1
        digit = Nth % pow if Nth % pow != 0 else pow
        
        #print "count, digit: ", count, digit
        N = (10 ** (pow - 1) + count - 1)
        n = pow - digit
        #print N
        while n > 0:
            N /= 10
            n -= 1
        return N % 10 

sol = Solution()
print sol.findNthDigit(15)


